File: corpus.py
"""Build a fully-expanded, self-contained corpus file from a study shortlist.

The harvest writes *summaries*: accession, title, abstract, record count, and
the classified publications. Everything a model layer actually reads — the
per-sample attribute bags, the experiment and run rows, the paper text — is
re-fetched from SRA and Europe PMC at reconstruction time, on every run.

That is fine for one run and wrong for everything else:

* **Reproducibility.** Two runs a week apart are not the same experiment. A
  model comparison across them mixes the model's effect with whatever the
  archives changed in between.
* **A language boundary.** Anything that consumes reconstruction input has to
  speak SRA, Europe PMC and Unpaywall. With a corpus file it has to speak JSON.
* **Auditability.** :mod:`audit` cannot check layer 4 at all today, for the
  single reason that paper text is never persisted. Persisting it — measured at
  30 KB per paper, 9 MB for the whole reference corpus — closes that hole.

    import corpus
    corpus.build_full_corpus("datasets/oa_corpus.json",
                             "datasets/oa_corpus_full.json")

Costs no model tokens. It is NCBI, Europe PMC and Unpaywall only, all free, and
on the 346-study reference corpus it runs for roughly two and a half hours —
almost all of it waiting on SRA to return 102,227 records. Checkpointed every
``checkpoint_every`` studies so an interrupted build resumes instead of
restarting.

FILE SHAPE
``papers`` is keyed by publication id; each publication's ``paper_ids`` indexes
into it. Shared rather than inlined because 17 papers serve more than one study,
and a paper can then be re-fetched without rebuilding a study blob. A list per
publication, so a study with several open-access papers keeps all of them and a
second copy of one article can be added later without a format change.
"""

from __future__ import annotations

from datetime import datetime, timezone
import json
import os

from project import (Project, classify_publication, fetch_biosample_records,
                     fetch_bioproject_record, fetch_open_access_text,
                     publication_oa_status, load_studies)
import dataset
import reconstruct

# Bumped when the file shape changes in a way a reader must notice. Readers
# should refuse a version they do not know rather than guess — a corpus built
# a year from now will not be this one.
FORMAT_VERSION = 2

# Studies between checkpoint writes. Each study is minutes of fetching, so a
# small number costs little and saves a lot when a build is interrupted.
CHECKPOINT_EVERY = 10


def _checkpoint_path(path):
    return f"{path}.partial"


def _classify_all(publications, sleep=None):
    """Classify every publication, not just up to the first open-access one.

    :func:`dataset.filter_oa_studies` deliberately stops at the first ``oa``
    hit — for deciding whether a study qualifies, the rest cannot change the
    answer. The corpus is not making that decision, and inheriting the
    short-circuit left **80 of 432 publications unclassified** in the previous
    build: 50 studies have more than one paper and nobody had checked whether
    the extras were retrievable.

    Returns the number newly classified. ``accessibility_type`` stays None only
    if the lookup itself fails, which keeps "never checked" distinguishable from
    "checked, not open".
    """
    done = 0
    for pub in publications:
        if pub.accessibility_type is None:
            try:
                pub.accessibility_type = classify_publication(pub.id, sleep=sleep)
                done += 1
            except Exception:      # noqa: BLE001 - one bad id must not stop a build
                pass
    return done


def _oa_publications(study):
    """Every publication classified ``oa``, in order.

    All of them, not just the first: layer 4 may choose to read only one, but
    that is a *reconstruction* policy and baking it into the corpus would mean a
    rebuild to revisit it.
    """
    return [p for p in study.publications if p.accessibility_type == "oa"]


def build_full_corpus(
    in_path="datasets/oa_corpus.json",
    out_path="datasets/oa_corpus_full.json",
    max_records=dataset.MAX_RECORDS,
    biosample=True,
    bioproject=True,
    papers=True,
    resume=True,
    checkpoint_every=CHECKPOINT_EVERY,
    limit=None,
):
    """Expand every study in ``in_path`` and write a self-contained corpus.

    ``biosample`` additionally fetches each sample's BioSample record and keeps
    it in :attr:`project.Sample.biosample_attributes`, *beside* the SRA bag
    rather than merged into it — the two disagree often enough to be worth
    telling apart, and merging would change what layer 3 gets credited with
    inferring. Batched, so it adds minutes rather than hours.

    ``papers`` fetches the full text of each study's first open-access
    publication into the shared ``papers`` map. Text is capped by
    :func:`project.fetch_open_access_text` at
    :data:`reconstruct.PAPER_MAX_CHARS`, which is what layer 4 would have seen.

    ``limit`` builds only the first N studies — for trying the shape out before
    committing to the full run.

    A study that fails to expand is kept in summary form with a ``build_note``
    explaining why, exactly as :func:`dataset._expand` does mid-pipeline: one
    unreachable study must not cost the other 345.
    """
    studies = load_studies(in_path)
    if limit is not None:
        studies = studies[:limit]

    ckpt = _checkpoint_path(out_path)
    params = {"source": str(in_path), "max_records": max_records,
              "biosample": bool(biosample), "bioproject": bool(bioproject),
              "papers": bool(papers),
              "study_count": len(studies), "format_version": FORMAT_VERSION}
    state = dataset._load_checkpoint(ckpt, params) if resume else None
    done = state["done"] if state else {}
    paper_texts = state.get("papers", {}) if state else {}
    if state:
        print(f"Resuming {ckpt}: {len(done)} of {len(studies)} studies already built")
    else:
        print(f"Building {len(studies)} studies -> {out_path}")

    notes = 0
    for n, study in enumerate(studies, 1):
        if study.accession in done:
            continue
        full, note = dataset._expand(study, max_records)
        if note:
            notes += 1
        # _expand deliberately drops publications (it skips the BioProject
        # fetch); they carry the `oa` classes stage 2 paid for, so restore them.
        full.publications = study.publications

        n_bio = 0
        if biosample and not note:
            ids = [s.biosample for s in full.samples.values() if s.biosample]
            if ids:
                try:
                    got = fetch_biosample_records(ids)
                    for sample in full.samples.values():
                        entry = got.get(sample.biosample or "")
                        if entry:
                            sample.biosample_attributes, sample.biosample_record = entry
                            n_bio += 1
                except Exception as exc:      # noqa: BLE001 - never fail the build
                    note += f" [biosample: {type(exc).__name__}: {exc}]"

        # The BioProject record proper. `_expand` skips it to save a fetch per
        # study, so the publications restored above carry only what stage 2
        # classified — this adds the project's own description, target and dates.
        if bioproject and not note and full.bioproject:
            try:
                got = fetch_bioproject_record(full.bioproject)
                if got:
                    full.bioproject_record, bp_pubs = got
                    # Merge the BioProject's own publication metadata (date,
                    # status, reference) into the classified list, matched by id.
                    by_id = {p.id: p for p in bp_pubs}
                    for pub in full.publications:
                        extra = by_id.get(pub.id)
                        if extra:
                            pub.date = pub.date or extra.date
                            pub.status = pub.status or extra.status
                            pub.reference = pub.reference or extra.reference
            except Exception as exc:          # noqa: BLE001
                note += f" [bioproject: {type(exc).__name__}: {exc}]"

        # Classify the whole bibliography before choosing what to fetch, or the
        # short-circuit in stage 2 silently caps this study at one paper.
        n_new = _classify_all(full.publications) if papers else 0

        for pub in (_oa_publications(full) if papers else []):
            if pub.id not in paper_texts:
                try:
                    text = fetch_open_access_text(pub.id)
                except Exception as exc:      # noqa: BLE001
                    text = None
                    note += f" [paper {pub.id}: {type(exc).__name__}: {exc}]"
                # An `oa` paper whose text will not come back is normal, not an
                # error: bronze and green are genuinely open access and still
                # deposit nothing to Europe PMC, the only source the fetch reads.
                try:
                    oa_status = publication_oa_status(pub.id)
                except Exception:             # noqa: BLE001 - never fail the build
                    oa_status = None
                paper_texts[pub.id] = {
                    "id": pub.id, "type": pub.type,
                    "chars": len(text) if text else 0,
                    "text": text or None,
                    "oa_status": oa_status,
                    # 91% of stored papers hit the cap; flag it so nothing
                    # downstream treats a fragment as a whole document.
                    "truncated": bool(text) and
                                 len(text) >= reconstruct.PAPER_MAX_CHARS,
                }
            # Point the publication at its text. Keyed by publication id today;
            # a list so a second copy of the same article can be added later
            # without changing the format.
            if paper_texts.get(pub.id, {}).get("text"):
                pub.paper_ids = [pub.id]

        blob = full.to_dict()
        blob["build_note"] = note.strip() or None
        done[study.accession] = blob

        print(f"  [{n}/{len(studies)}] {study.accession}: "
              f"{len(full.samples):,} samples, {len(full.experiments):,} experiments"
              f"{f', {n_bio} biosample' if n_bio else ''}"
              f"{f', +{n_new} classified' if n_new else ''}"
              f"{'  ' + note.strip() if note else ''}", flush=True)

        if len(done) % checkpoint_every == 0:
            dataset._write_json(ckpt, {"version": dataset._CHECKPOINT_VERSION,
                                       "params": params, "done": done,
                                       "papers": paper_texts})

    payload = {
        "format_version": FORMAT_VERSION,
        "created": datetime.now(timezone.utc).isoformat(timespec="seconds"),
        "params": params,
        "counts": {
            "studies": len(done),
            "records": sum(len(b.get("experiments") or []) for b in done.values()),
            "samples": sum(len(b.get("samples") or {}) for b in done.values()),
            "papers": sum(1 for p in paper_texts.values() if p["text"]),
            "papers_empty": sum(1 for p in paper_texts.values() if not p["text"]),
            "summary_only": notes,
        },
        "papers": paper_texts,
        "studies": [done[s.accession] for s in studies if s.accession in done],
    }
    dataset._write_json(out_path, payload)
    if os.path.exists(ckpt):
        os.remove(ckpt)

    c = payload["counts"]
    print(f"\nWrote {out_path}  ({os.path.getsize(out_path)/1e6:.1f} MB)")
    print(f"  studies {c['studies']:,} | samples {c['samples']:,} | "
          f"experiments {c['records']:,}")
    print(f"  papers {c['papers']:,} with text, {c['papers_empty']:,} classified oa "
          f"but empty | {c['summary_only']:,} studies summary-only")
    return payload


def load_full_corpus(path):
    """Read a corpus file back as ``(projects, papers, meta)`` — no network.

    Refuses a ``format_version`` it does not know rather than guessing at the
    shape: a reader that silently mis-parses a future corpus is worse than one
    that stops.

    ``build_note`` is re-attached to each project;
    :meth:`project.Project.from_dict` only reads the keys it declares.

    The link from a study to its text lives on each publication's
    ``paper_ids``, which index into the returned ``papers`` map. A study can
    therefore carry several papers, or none, without a format change.
    """
    with open(path, encoding="utf-8") as fh:
        payload = json.load(fh)
    version = payload.get("format_version")
    if version != FORMAT_VERSION:
        raise ValueError(
            f"{path} is format_version {version!r}; this build understands "
            f"{FORMAT_VERSION}. Rebuild it, or read it with the matching version."
        )
    projects = []
    for blob in payload["studies"]:
        p = Project.from_dict(blob)
        p.build_note = blob.get("build_note")
        projects.append(p)
    return projects, payload["papers"], {k: v for k, v in payload.items()
                                         if k not in ("studies", "papers")}


def paper_texts(project, papers) -> list[str]:
    """Every retrieved text for this study, in publication order. Often empty.

    Empty covers two cases a caller usually treats alike: no open-access
    publication at all, and one that is genuinely open but whose text could not
    be retrieved. See :func:`split_by_paper` for why the second happens.
    """
    out = []
    for pub in project.publications:
        for pid in getattr(pub, "paper_ids", None) or []:
            text = (papers.get(pid) or {}).get("text")
            if text:
                out.append(text)
    return out


def paper_text(project, papers) -> str | None:
    """The first retrieved text, or None — what layer 4 actually reads today."""
    texts = paper_texts(project, papers)
    return texts[0] if texts else None


def split_by_paper(projects, papers):
    """``(with_text, without_text)`` — the split layer 4 actually cares about.

    **A study without usable paper text does not look "bad" in the data.** Its
    ``publications`` list is populated, the publication's
    ``accessibility_type`` is ``"oa"``, and ``build_note`` is None — because
    none of that is wrong. The paper genuinely is open
    access; it is open by a route (*bronze* at the publisher, or *green* in a
    repository) that deposits nothing to Europe PMC, which is the only source
    :func:`project.fetch_open_access_text` reads.

    So filtering on publications or on ``accessibility_type`` will not find
    them. The only signal is an empty ``paper_ids``, or an empty ``text`` behind
    one, which is what this checks.
    """
    with_text, without = [], []
    for p in projects:
        (with_text if paper_text(p, papers) else without).append(p)
    return with_text, without


if __name__ == "__main__":
    import sys
    from project import set_entrez_credentials

    def _cred(path):
        with open(path, encoding="utf-8") as fh:
            return fh.read().strip()

    set_entrez_credentials(api_key=_cred("api_key.txt"), email=_cred("email.txt"))
    limit = int(sys.argv[1]) if len(sys.argv) > 1 else None
    build_full_corpus(limit=limit)

File: rust/src/corpus.rs
use serde::{Deserialize, Serialize};
use std::collections::BTreeMap;

use crate::dto::CorpusDto;
use crate::project::{Paper, Project, ZonedDate};

#[derive(Clone, Debug, Serialize, Deserialize)]
pub struct Corpus {
    pub format_version: u32,
    pub created: ZonedDate,
    pub params: CorpusParams,
    pub counts: CorpusCounts,
    pub projects: Vec<Project>,
}

// The version this build understands. Refusing an unknown version beats
// mis-parsing one: the corpus format has already changed once (v1 carried no
// submission, BioProject record, run statistics or per-publication papers).
pub const SUPPORTED_FORMAT_VERSION: u32 = 2;

#[derive(Debug)]
pub enum CorpusError {
    Json(serde_json::Error),
    UnsupportedVersion { found: u32, supported: u32 },
}

impl std::fmt::Display for CorpusError {
    #[inline]
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        match self {
            CorpusError::Json(e) => write!(f, "corpus JSON is malformed: {e}"),
            CorpusError::UnsupportedVersion { found, supported } => write!(
                f,
                "corpus is format_version {found}; this build understands \
                 {supported}. Rebuild it, or read it with the matching version."
            ),
        }
    }
}

impl std::error::Error for CorpusError {}

impl From<serde_json::Error> for CorpusError {
    #[inline]
    fn from(e: serde_json::Error) -> Self {
        CorpusError::Json(e)
    }
}

impl Corpus {
    // Two passes by necessity: publications reference papers by id, so the map
    // has to be built before any study can resolve its own. It is local to this
    // function — once every publication owns its paper the index is redundant,
    // and keeping it would duplicate ~353 texts (roughly 10 MB) for nothing.
    #[inline]
    pub fn from_json(json: &str, print: bool) -> Result<Self, CorpusError> {
        let dto: CorpusDto = serde_json::from_str(json)?;
        if dto.format_version != SUPPORTED_FORMAT_VERSION {
            return Err(CorpusError::UnsupportedVersion {
                found: dto.format_version,
                supported: SUPPORTED_FORMAT_VERSION,
            });
        }

        let papers: BTreeMap<String, Paper> = dto
            .papers
            .into_iter()
            .map(|(id, p)| (id, Paper::from(p)))
            .collect();

        let projects = dto
            .studies
            .into_iter()
            .map(|s| s.into_project(&papers, dto.format_version, &dto.created))
            .collect();

        let corpus = Corpus {
            format_version: dto.format_version,
            created: ZonedDate { raw: dto.created },
            params: dto.params,
            counts: dto.counts,
            projects,
        };

        if print {
            println!("format_version {}", corpus.format_version);
            println!("projects       {}", corpus.projects.len());
            println!("with text      {}", corpus.with_paper_text().count());

            let owned: usize = corpus.projects.iter().map(|p| papers_of(p).len()).sum();
            let multi: Vec<_> = corpus
                .projects
                .iter()
                .filter(|p| papers_of(p).len() > 1)
                .collect();
            println!("papers owned by publications {owned}");
            println!("studies with >1 paper        {}", multi.len());

            if let Some(p) = multi.first() {
                println!("\n{:?}", p.accession);
                for pub_ in &p.publications {
                    match &pub_.paper {
                        Some(paper) => println!(
                            "   {} {:?} -> {} chars, truncated={}, oa_status={:?}",
                            pub_.id,
                            pub_.accessibility_type,
                            paper.char_count,
                            paper.truncated,
                            paper.oa_status
                        ),
                        None => println!("   {} {:?} -> no paper", pub_.id, pub_.accessibility_type),
                    }
                }
            }

            let s = &corpus.projects[0];
            let sample = s.samples.values().next().unwrap();
            println!("\n{:?} sample {:?}", s.accession, sample.accession);
            println!("   sra attrs       {}", sample.attributes.len());
            println!("   biosample attrs {}", sample.biosample_attributes.len());
            println!("   biosample pkg   {:?}",
                    sample.biosample.as_ref().and_then(|b| b.package.as_ref()));
        }

        Ok(corpus)
    }

    // Studies whose paper text actually came back. A study without it does not
    // look wrong in the data — its publications are populated and classified
    // `oa` — because nothing is wrong: bronze and green routes are genuinely
    // open access and still deposit nothing to Europe PMC, which is the only
    // source the harvest reads. The sole signal is absent or empty text.
    #[inline]
    pub fn with_paper_text(&self) -> impl Iterator<Item = &Project> {
        self.projects.iter().filter(|p| !papers_of(p).is_empty())
    }
}

#[derive(Clone, Debug, Serialize, Deserialize)]
pub struct CorpusParams {
    pub source: String,
    pub max_records: usize,
    pub biosample: bool,
    pub bioproject: bool,
    pub papers: bool,
    pub study_count: usize,
    pub format_version: u32,
}

#[derive(Clone, Debug, Serialize, Deserialize)]
pub struct CorpusCounts {
    pub studies: usize,
    pub records: usize,
    pub samples: usize,
    pub papers: usize,
    pub papers_empty: usize,
    pub summary_only: usize,
}

// Every retrieved text a study actually has. 37 of 346 studies carry more than
// one, which is why this is a Vec and not an Option — the short-circuit that
// used to cap studies at a single paper was removed from the harvest.
#[inline]
pub fn papers_of(project: &Project) -> Vec<&Paper> {
    project
        .publications
        .iter()
        .filter_map(|p| p.paper.as_ref())
        .filter(|p| p.text.is_some())
        .collect()
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::project::PublicationDb;

    // A real slice of `datasets/oa_corpus_full.json`, not a hand-written file:
    // the point of the fixture is to prove the deserializer matches what the
    // builder actually emits, and a fixture written from the DTO definitions
    // would only prove it matches itself. Trimmed to 3 studies / 4 samples and
    // paper texts cut to 280 chars, with `chars` corrected to match.
    //
    // Three studies, chosen to cover each branch of the paper lookup:
    //   DRP006604  3 publications, all resolving to papers with text
    //   DRP003937  1 publication, `oa`, but no paper_ids at all - the bronze /
    //              green case, open access with nothing deposited to Europe PMC
    //   SRP999999  SYNTHETIC. A publication pointing at a paper whose text is
    //              empty. No study in the real corpus does this - all 61
    //              empty-text papers are orphans, referenced by nothing - but
    //              the text filter in `papers_of` is reachable code, so the
    //              case has to be manufactured to cover it. The accession is
    //              deliberately not a real one.
    const MINI: &str = include_str!("../test_data/mini_corpus.json");

    fn corpus() -> Corpus {
        Corpus::from_json(MINI, false).expect("fixture should parse")
    }

    fn project(accession: &str) -> Project {
        corpus()
            .projects
            .into_iter()
            .find(|p| p.accession.0 == accession)
            .unwrap_or_else(|| panic!("{accession} missing from fixture"))
    }

    #[test]
    fn test_from_json() {
        let corpus = corpus();

        assert_eq!(corpus.format_version, SUPPORTED_FORMAT_VERSION);
        assert_eq!(corpus.created.raw, "2026-08-16T04:42:51+00:00");
        assert_eq!(corpus.params.source, "datasets/oa_corpus.json");
        assert_eq!(corpus.counts.studies, 3);
        assert_eq!(corpus.projects.len(), corpus.counts.studies);

        // The two-pass paper resolution: publications carry `paper_ids` on the
        // wire and own a `Paper` afterwards. Checking the id alone would pass
        // even if the map lookup silently handed back the wrong entry, so the
        // text is checked too.
        let multi = project("DRP006604");
        assert_eq!(multi.publications.len(), 3);
        for publication in &multi.publications {
            let paper = publication.paper.as_ref().expect("paper should resolve");
            assert_eq!(paper.id, publication.id);
            assert_eq!(paper.char_count, 280);
            assert_eq!(paper.text.as_ref().map(|t| t.chars().count()), Some(280));
            assert!(paper.truncated);
        }

        // A publication with no paper_ids resolves to no paper, rather than to
        // an empty one.
        let no_ids = project("DRP003937");
        assert_eq!(no_ids.publications.len(), 1);
        assert!(no_ids.publications[0].paper.is_none());

        // The DOI publication proves `type` is parsed, not passed through: the
        // wire says "eDOI" and the other two say "ePubmed".
        let dois = multi
            .publications
            .iter()
            .filter(|p| p.db_type == PublicationDb::EDoi)
            .count();
        assert_eq!(dois, 1);

        // Nested structure survives the trim, so the fixture still exercises
        // the deep DTO conversions rather than just the top level.
        assert_eq!(multi.samples.len(), 2);
        assert_eq!(multi.experiments.len(), 2);
        assert!(multi.experiments.iter().all(|e| !e.runs.is_empty()));
        // Every experiment's sample_ids resolve; a fixture that lost this would
        // make the target-schema tests pass for the wrong reason.
        for experiment in &multi.experiments {
            for id in &experiment.sample_ids {
                assert!(multi.samples.contains_key(id), "dangling sample id {id:?}");
            }
        }

        // Corpus-format bookkeeping is copied onto every project, not just kept
        // at the root, so a Project separated from its Corpus still knows how
        // stale it is.
        assert_eq!(multi.source.corpus_format_version, SUPPORTED_FORMAT_VERSION);
        assert_eq!(multi.source.fetched_at.raw, corpus.created.raw);
    }

    #[test]
    fn test_from_json_rejects_unsupported_version() {
        // Only the first occurrence: `format_version` appears again inside
        // `params`, and the check reads the root one.
        let bumped = MINI.replacen(
            &format!("\"format_version\": {SUPPORTED_FORMAT_VERSION}"),
            "\"format_version\": 99",
            1,
        );
        match Corpus::from_json(&bumped, false) {
            Err(CorpusError::UnsupportedVersion { found, supported }) => {
                assert_eq!(found, 99);
                assert_eq!(supported, SUPPORTED_FORMAT_VERSION);
            }
            other => panic!("expected UnsupportedVersion, got {other:?}"),
        }
    }

    #[test]
    fn test_from_json_rejects_malformed() {
        // Truncation rather than garbage: a half-written corpus is the failure
        // that actually happens, and it is the one that could plausibly parse
        // into a short-but-valid-looking Corpus.
        let truncated = &MINI[..MINI.len() / 2];
        assert!(matches!(
            Corpus::from_json(truncated, false),
            Err(CorpusError::Json(_))
        ));
    }

    #[test]
    fn test_with_paper_text() {
        let corpus = corpus();
        let with_text: Vec<_> = corpus
            .with_paper_text()
            .map(|p| p.accession.0.as_str())
            .collect();

        // One of three. The other two are the reason this method exists: both
        // carry a publication classified `oa`, and neither yields any text.
        assert_eq!(with_text, vec!["DRP006604"]);

        // Having publications is not the same as having text, which is the
        // distinction the whole method turns on.
        assert!(corpus.projects.iter().all(|p| !p.publications.is_empty()));
    }

    #[test]
    fn test_papers_of() {
        // Every publication resolves, so this is 3 and not 1: the single-paper
        // short-circuit that used to cap a study at one paper is gone.
        assert_eq!(papers_of(&project("DRP006604")).len(), 3);

        // No paper_ids at all - nothing to filter, nothing to return.
        assert!(papers_of(&project("DRP003937")).is_empty());

        // The paper resolves and is attached, but its text is empty, so it is
        // filtered out. This is the branch the synthetic study exists for.
        let empty = project("SRP999999");
        assert!(empty.publications[0].paper.is_some());
        assert!(papers_of(&empty).is_empty());

        // Order follows the publication order, which the model's evidence
        // string depends on: a set here would reintroduce the nondeterminism
        // the BTreeMaps were chosen to avoid.
        let multi = project("DRP006604");
        let ids: Vec<_> = papers_of(&multi).iter().map(|p| p.id.as_str()).collect();
        assert_eq!(ids, vec!["10.1038/s41598-020-72888-6", "32973264", "36223455"]);
    }
}